### ISSUE TITLE
Batch lookup tool for location geodata

### ISSUE THREAD

**Issue by densho** [denshoproject/namesdb-editor#38]
The Person-Locations table contains a text label for location names; but also includes the ability to store latitude and longitude coordinates. This lat-long point data will be needed for the mini-map visualization that is planned for the public NR interface. 

We need a tool (and/or workflow) that can feed location names to a batch geodata resolver service such as Nominatim, and then produce a datafile that can be used to batch update Person-Locations records in the namesdb. 

We can get the initial list of location names from the `FARRecord` and `WRARecord` tables, which can then be deduplicated to reduce the size of the batch operation.

**Comment by densho** [comment]
related to #39

**Comment by densho** [comment]
See also: https://github.com/denshoproject/sites-of-shame-app/tree/main/data-processing

**Comment by gjost** [comment]
Geocod.io (used by Stamen in SoS) is a paid service where you upload CSV files and get back the CSV with address data in additional columns.  There's an API with a third-party Python library.

We currently have 188260 `PersonFacility` records.  At Geocod.io's pricing of 0.50/1000 (https://www.geocod.io/pricing/) it would cost ~$94.

**Comment by gjost** [comment]
Nominatim has an undocumented batch search mode but it appears to be broken: https://nominatim.org/release-docs/latest/customize/Settings/#nominatim_search_batch_mode

**Comment by GeoffFroh** [comment]
> Geocod.io (used by Stamen in SoS) is a paid service where you upload CSV files and get back the CSV with address data in additional columns. There's an API with a third-party Python library.
> 
> We currently have 188260 `PersonFacility` records. At Geocod.io's pricing of 0.50/1000 (https://www.geocod.io/pricing/) it would cost ~$94.

What if we were to dedupe the locations? That would reduce the number substantially, I'd wager. 

And we also already have geodata for the facilities so no need to look those up since we'll just be adding those locations to the new `Locations` table from the `Facilities` table

**Comment by gjost** [comment]
Yep indeed - I'll have to see how many unique values are in there

**Comment by gjost** [comment]
I've played with both Nominatim and Geocod.io.

Nominatim
- Doesn't really have a batch mode, or rather it sortof says it does but it's broken.
- search/reverse output in various formats, including `geocodejson`

Geocod.io
- I wrote a function for doing reverse geocode search (lat,lng to address) for multiple `Location` objects.
- Returns multiple possible results i.e. addresses for each coordinate, with varying degrees of accuracy.
- Each result has address as a single string and as a dict of components.
- In my tests it returned results from multiple sources incl `TIGER/Line® dataset from the US Census Bureau`

Some questions
- How do we decide which result/address to use?
- I think we might want to keep the formatted address and a JSON string of the address components.  Nominatim and Geocod.io use different component keys.  `geocodejson` seems like it might be a standard for key naming?
- Maybe we want to keep *all* the Geocod.io results for each `Location`?

**Comment by GeoffFroh** [comment]
> Nominatim
> 
> * Doesn't really have a batch mode, or rather it sortof says it does but it's broken.
> * search/reverse output in various formats, including `geocodejson`

I don't think it's worth putting additional time into Nominatum for the batch tool. It's possible that we could use it for one-off lookup after we've done this large initial batch.

> Geocod.io
> 
> * I wrote a function for doing reverse geocode search (lat,lng to address) for multiple `Location` objects.
> * Returns multiple possible results i.e. addresses for each coordinate, with varying degrees of accuracy.
> * Each result has address as a single string and as a dict of components.
> * In my tests it returned results from multiple sources incl `TIGER/Line® dataset from the US Census Bureau`

Remember that we're actually trying to do the opposite operation here: feed the API a string location/address, then receive the lat/lng point data.

**Comment by gjost** [comment]
> Remember that we're actually trying to do the opposite operation here: feed the API a string location/address, then receive the lat/lng point data.

That will probably make things easier!

**Comment by gjost** [comment]
Notes: 

- The original `facilities.xml` just has city/state and lat/lng, with no street addresses or postalcodes.
- `facilities.xml` does have a TGN ID for each(?) facility.
- FAR and WRA records didn't have full addresses, just city/state.

Reverse-geocoding several facilities may not give an address (Manzanar's lat/lng just returned something like "Highway 395, Inyo County, CA".  (I know we're not reverse-geocoding, just an observation).

**Comment by gjost** [comment]
FAR records have camp block, barrack, and room fields, fwiw
